File: collector/src/Vincere.AutoExport.NinjaTrader.Core/Capture/OrderAttribution.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Vincere.AutoExport.NinjaTrader.Core.Capture
{

/* ---------------------------------------------------------------------------
 * WHICH ALGORITHM PLACED THE ORDERS NOBODY KEPT A RECORD FOR.
 *
 * The ledger beside this holds every link the platform asserted while we were
 * watching. Everything older than the agent is unattributed and cannot be
 * recovered from NinjaTrader, because the cascade already took it.
 *
 * It can be INFERRED, and the measurement that says so came off a real VPS on
 * 2026-09-24. Three strategies still had live links there, and the three
 * signatures do not overlap:
 *
 *   G4M-3.4        MES  PT1/PT2/PT3   enters 4, exits 2/1/1
 *   URGO-4.5       MNQ  PT1/PT2/PT3   enters 2 or 4
 *   Bullet Bot-1.1 NQ   a single PT   flat 2
 *
 * Across the whole seven months those axes keep separating: nine instruments,
 * and the exit ladder falls into four families - a lone PT (50 account and
 * instrument pairs), PT1 (125), PT1+PT2 (81), PT1+PT2+PT3 (46).
 *
 * THE ONE RULE THIS FILE EXISTS TO OBEY. A wrong strategy name is worse than
 * no strategy name. Absent, an account day shows up as unattributable and
 * somebody looks at it. Wrong, it silently moves a day's losses onto an
 * algorithm that never traded them, and every number computed from it after
 * that is confidently false. So:
 *
 *   - a signature that matches two strategies attributes to NEITHER;
 *   - a signature learned from a single order is not a signature;
 *   - every inferred row is labelled `inferred` and can be dropped by any
 *     reader that only wants record.
 *
 * WHAT IT CANNOT DO, stated so nobody has to discover it. It answers the algo,
 * not the version: URGO-4.5 and URGO-4.4 on the same instrument with the same
 * ladder are one signature. And an algorithm retired before the agent arrived
 * has no example to learn from, so its orders stay unattributed forever. That
 * is the honest outcome and the ledger is the only cure, going forward.
 * ------------------------------------------------------------------------- */

/// <summary>An order as the attribution needs to see it.</summary>
public sealed class AttributableOrder
{
    public AttributableOrder(long orderId, string accountName, string instrument, string orderName, int quantity)
    {
        OrderId = orderId;
        AccountName = accountName;
        Instrument = instrument;
        OrderName = orderName;
        Quantity = quantity;
    }

    public long OrderId { get; private set; }
    public string AccountName { get; private set; }
    public string Instrument { get; private set; }
    public string OrderName { get; private set; }
    public int Quantity { get; private set; }
}

public enum AttributionBasis
{
    /// <summary>The platform said so and we wrote it down before the cascade.</summary>
    Record,
    /// <summary>Nothing said so; the signature matched exactly one strategy.</summary>
    Inferred,
    /// <summary>Nothing said so and the signature did not settle it.</summary>
    None,
}

public sealed class AttributedOrder
{
    public AttributedOrder(long orderId, string strategyName, AttributionBasis basis, string signature)
    {
        OrderId = orderId;
        StrategyName = strategyName;
        Basis = basis;
        Signature = signature;
    }

    public long OrderId { get; private set; }
    public string StrategyName { get; private set; }
    public AttributionBasis Basis { get; private set; }
    public string Signature { get; private set; }
}

/// <summary>
/// What an order looks like to the matcher: the instrument it traded, the shape
/// of the exit ladder its strategy uses, and the size that ladder starts from.
///
/// The order NAME is in the signature only as its family (`PT1` from
/// `PT1-Short`), never its side. A strategy that goes long on Monday and short
/// on Tuesday is one strategy, and folding the side in would make it two.
/// </summary>
public sealed class OrderSignature : IEquatable<OrderSignature>
{
    public OrderSignature(string instrument, string ladder, int entryQuantity)
    {
        Instrument = instrument ?? string.Empty;
        Ladder = ladder ?? string.Empty;
        EntryQuantity = entryQuantity;
    }

    public string Instrument { get; private set; }
    public string Ladder { get; private set; }
    public int EntryQuantity { get; private set; }

    public bool Equals(OrderSignature other)
    {
        return other != null
            && string.Equals(Instrument, other.Instrument, StringComparison.OrdinalIgnoreCase)
            && string.Equals(Ladder, other.Ladder, StringComparison.OrdinalIgnoreCase)
            && EntryQuantity == other.EntryQuantity;
    }

    public override bool Equals(object obj) { return Equals(obj as OrderSignature); }

    public override int GetHashCode()
    {
        unchecked
        {
            int hash = 17;
            hash = (hash * 31) + StringComparer.OrdinalIgnoreCase.GetHashCode(Instrument);
            hash = (hash * 31) + StringComparer.OrdinalIgnoreCase.GetHashCode(Ladder);
            hash = (hash * 31) + EntryQuantity;
            return hash;
        }
    }

    public override string ToString() { return Instrument + "|" + Ladder + "|" + EntryQuantity; }
}

public static class OrderAttribution
{
    /// <summary>A ladder learned from fewer than this many orders is a coincidence.</summary>
    public const int MinimumOrdersPerSignature = 4;

    /// <summary>`PT1-Short` and `PT1-Long` are the same rung.</summary>
    public static string NameFamily(string orderName)
    {
        string name = (orderName ?? string.Empty).Trim();
        if (name.Length == 0) return string.Empty;
        int dash = name.IndexOf('-');
        string head = dash > 0 ? name.Substring(0, dash) : name;
        // "Enter Long" / "Stop Short" split on the space instead.
        int space = head.IndexOf(' ');
        if (space > 0) head = head.Substring(0, space);
        return head.Trim();
    }

    /// <summary>
    /// The exit ladder of a group of orders, as a stable string.
    ///
    /// Only the profit-target rungs, sorted: entries and stops are present in
    /// every strategy here and separate nothing. `PT` alone is a different
    /// shape from `PT1`, which is why the family is kept verbatim rather than
    /// stripped of its digit.
    /// </summary>
    public static string Ladder(IEnumerable<AttributableOrder> orders)
    {
        var rungs = (orders ?? Enumerable.Empty<AttributableOrder>())
            .Select(order => NameFamily(order?.OrderName))
            .Where(family => family.StartsWith("PT", StringComparison.OrdinalIgnoreCase))
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .OrderBy(family => family, StringComparer.OrdinalIgnoreCase)
            .ToList();
        return rungs.Count == 0 ? string.Empty : string.Join("+", rungs);
    }

    /// <summary>The size the strategy enters with, which the exits scale out of.</summary>
    public static int EntryQuantity(IEnumerable<AttributableOrder> orders)
    {
        var entries = (orders ?? Enumerable.Empty<AttributableOrder>())
            .Where(order => order != null
                && NameFamily(order.OrderName).Equals("Enter", StringComparison.OrdinalIgnoreCase)
                && order.Quantity > 0)
            .Select(order => order.Quantity)
            .ToList();
        if (entries.Count == 0) return 0;
        // The mode, not the maximum: a strategy that usually enters 2 and once
        // entered 4 on a re-entry is a 2.
        return entries.GroupBy(quantity => quantity)
            .OrderByDescending(group => group.Count())
            .ThenByDescending(group => group.Key)
            .First().Key;
    }

    /// <summary>
    /// Learn one signature per strategy from the orders we KNOW belong to it.
    ///
    /// Grouped by strategy and instrument, because the same algorithm on MES
    /// and on MNQ is two signatures and must not be averaged into one that
    /// matches neither.
    /// </summary>
    public static IReadOnlyDictionary<OrderSignature, string> LearnSignatures(
        IEnumerable<AttributableOrder> orders,
        IReadOnlyDictionary<long, string> strategyByOrderId)
    {
        var byStrategyAndInstrument = new Dictionary<Tuple<string, string>, List<AttributableOrder>>();
        foreach (AttributableOrder order in orders ?? Enumerable.Empty<AttributableOrder>())
        {
            if (order is null) continue;
            if (strategyByOrderId is null) break;
            string strategy;
            if (!strategyByOrderId.TryGetValue(order.OrderId, out strategy)) continue;
            if (string.IsNullOrWhiteSpace(strategy)) continue;
            var key = Tuple.Create(strategy, order.Instrument ?? string.Empty);
            List<AttributableOrder> bucket;
            if (!byStrategyAndInstrument.TryGetValue(key, out bucket))
            {
                bucket = new List<AttributableOrder>();
                byStrategyAndInstrument[key] = bucket;
            }
            bucket.Add(order);
        }

        // A signature two strategies produce identifies neither, so it is
        // dropped rather than given to whichever was learned first.
        var claims = new Dictionary<OrderSignature, HashSet<string>>();
        foreach (KeyValuePair<Tuple<string, string>, List<AttributableOrder>> pair in byStrategyAndInstrument)
        {
            string strategy = pair.Key.Item1;
            string instrument = pair.Key.Item2;
            List<AttributableOrder> bucket = pair.Value;
            if (bucket.Count < MinimumOrdersPerSignature) continue;
            string ladder = Ladder(bucket);
            if (ladder.Length == 0) continue;
            var signature = new OrderSignature(instrument, ladder, EntryQuantity(bucket));
            HashSet<string> owners;
            if (!claims.TryGetValue(signature, out owners))
            {
                owners = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
                claims[signature] = owners;
            }
            owners.Add(strategy);
        }

        return claims
            .Where(pair => pair.Value.Count == 1)
            .ToDictionary(pair => pair.Key, pair => pair.Value.Single());
    }

    /// <summary>
    /// Attribute every order: from the ledger where we have it, from the
    /// signature where it settles, and from nothing otherwise.
    ///
    /// Orders are grouped by account and instrument before matching, because a
    /// ladder is a property of a run of trading and not of one order. One order
    /// on its own carries no ladder and is left alone.
    /// </summary>
    public static IReadOnlyList<AttributedOrder> Attribute(
        IEnumerable<AttributableOrder> orders,
        IReadOnlyDictionary<long, string> strategyByOrderId,
        IReadOnlyDictionary<OrderSignature, string> signatures)
    {
        var all = (orders ?? Enumerable.Empty<AttributableOrder>()).Where(order => order != null).ToList();
        var known = strategyByOrderId ?? new Dictionary<long, string>();
        var learned = signatures ?? new Dictionary<OrderSignature, string>();
        var results = new List<AttributedOrder>(all.Count);

        foreach (IGrouping<Tuple<string, string>, AttributableOrder> group in all.GroupBy(
            order => Tuple.Create(order.AccountName ?? string.Empty, order.Instrument ?? string.Empty)))
        {
            var bucket = group.ToList();
            string ladder = Ladder(bucket);
            int entry = EntryQuantity(bucket);
            var signature = new OrderSignature(group.Key.Item2, ladder, entry);
            string match;
            string inferred = ladder.Length > 0 && learned.TryGetValue(signature, out match) ? match : null;

            foreach (AttributableOrder order in bucket)
            {
                string recorded;
                if (known.TryGetValue(order.OrderId, out recorded) && !string.IsNullOrWhiteSpace(recorded))
                {
                    results.Add(new AttributedOrder(order.OrderId, recorded, AttributionBasis.Record, signature.ToString()));
                    continue;
                }
                results.Add(inferred == null
                    ? new AttributedOrder(order.OrderId, null, AttributionBasis.None, signature.ToString())
                    : new AttributedOrder(order.OrderId, inferred, AttributionBasis.Inferred, signature.ToString()));
            }
        }

        return results.OrderBy(result => result.OrderId).ToList();
    }
}

}

File: src/domain/orderAttribution.js
/* ---------------------------------------------------------------------------
 * WHICH ALGORITHM PLACED AN ORDER, WORKED OUT FROM THE SHAPE OF THE TRADE.
 *
 * This is the collector's attribution engine, on the desk's side of the wire.
 * The agent runs it on one machine against that machine's own database; this
 * runs it against the 59,264 orders the CRM already holds for every client, so
 * the answer does not wait for somebody to take a Deep Export.
 *
 * WHY IT IS NEEDED AT ALL. NinjaTrader records which strategy placed an order
 * and deletes that record by cascade the moment the strategy leaves the
 * workspace. Measured on a real VPS: 29 surviving links against 18,827 orders
 * over seven months. The desk rotates algorithms constantly, so it destroys its
 * own attribution constantly.
 *
 * HOW IT WORKS. A strategy template declares its geometry - how many contracts
 * each scale-out leg takes, and how far the stop and each profit target sit
 * from the entry in ticks - and a placed trade leaves exactly that geometry in
 * the order book. The G4M template declares 2/1/1 with stop 80 and targets
 * 80/120/160; its orders sat at exactly those distances. Tick for tick.
 *
 * THE ONE RULE. A wrong algorithm name is worse than none: absent, an account
 * day shows up as unattributable and somebody looks at it; wrong, it silently
 * moves a day's losses onto an algorithm that never traded them. So a geometry
 * two algorithms share attributes to neither, and a version two versions share
 * answers the algorithm with a null version rather than picking one.
 *
 * WHAT DIFFERS FROM THE AGENT'S COPY, and it is not the rule. The agent reads
 * NinjaTrader's own tables, so it knows each order's original price before a
 * trailing stop moved it, and the instrument's tick size. The CRM's `orders`
 * rows carry the price as submitted and no tick size at all, so the distances
 * come from TICK_SIZES below and the comparison is on what the rows can
 * support. Where that is not enough, this answers nothing.
 * ------------------------------------------------------------------------- */

/**
 * Tick size per instrument root, because `orders` does not carry one.
 *
 * These are contract specifications, not preferences: MES and MNQ move in
 * quarter points, YM in whole points, crude in cents. An instrument absent
 * here is not guessed at - its orders are left unattributed, which is the same
 * answer this gives for anything else it cannot measure.
 */
export const TICK_SIZES = Object.freeze({
  MES: 0.25, MNQ: 0.25, M2K: 0.1, MYM: 1,
  ES: 0.25, NQ: 0.25, RTY: 0.1, YM: 1,
  MGC: 0.1, GC: 0.1,
  CL: 0.01, MCL: 0.01,
  NG: 0.001,
  PL: 0.1,
});

/**
 * `MNQ 12-26` and `MNQ` are the same contract for this purpose.
 *
 * SPLIT ON THE SPACE, NOT ON THE DIGIT. A contract root can contain a digit -
 * M2K is the micro Russell - and cutting at the first one turns it into `M`,
 * which is in no tick table and silently drops the instrument. It cost 151
 * orders across 33 trades on the seven months measured. The expiry is always
 * separated by whitespace in this desk's data, in both forms it writes
 * (`MNQ 12-26` and `MNQ SEP26`).
 */
export function instrumentRoot(instrument) {
  const text = String(instrument || '').trim().toUpperCase();
  if (!text) return '';
  return text.split(/\s+/)[0] || '';
}

export function tickSizeFor(instrument) {
  return TICK_SIZES[instrumentRoot(instrument)] ?? null;
}

/**
 * The rung an order fills: `Enter`, `Stop`, `PT1`.
 *
 * The side is deliberately dropped. A strategy that goes long on Monday and
 * short on Tuesday is one strategy, and folding the side in would make it two.
 */
export function rungOf(orderName) {
  const name = String(orderName || '').trim();
  if (!name) return '';
  const head = name.split('-')[0];
  return head.split(' ')[0].trim();
}

export function isEntry(order) { return rungOf(order?.name).toLowerCase() === 'enter'; }
export function isStop(order) { return rungOf(order?.name).toLowerCase() === 'stop'; }
export function isTarget(order) { return rungOf(order?.name).toUpperCase().startsWith('PT'); }

/** `PT2-Short` is rung 2. A bare `PT` is rung 1. */
export function targetNumber(order) {
  if (!isTarget(order)) return 0;
  const digits = rungOf(order.name).replace(/\D/g, '');
  return digits ? Number(digits) : 1;
}

/**
 * Group one account's orders into trades: an entry and the exits belonging to
 * it.
 *
 * Two things break a naive walk and both are ordinary trading. A re-entry while
 * a position is open leaves the first trade's exits still live, so closing the
 * group on every entry hands them to the second. And an exit only belongs to an
 * entry if that entry has not already filled the same rung. So exits are given
 * to the OLDEST open entry with room for them, and one that fits nowhere is
 * dropped rather than attached to the nearest.
 */
export function reconstructTrades(orders = []) {
  const ordered = (orders || [])
    .filter(Boolean)
    .slice()
    .sort((a, b) => compare(a.time, b.time) || compare(a.id, b.id));

  const open = [];
  for (const order of ordered) {
    if (isEntry(order)) {
      open.push({ entry: order, exits: [] });
      continue;
    }
    const target = open.find((trade) => hasRoomFor(trade, order));
    if (target) target.exits.push(order);
  }
  return open;
}

/**
 * Order two of anything a caller might hand in as a time or an id.
 *
 * NUMBERS ARE COMPARED AS NUMBERS. Sorting them as text is the defect this
 * codebase has already paid for once: `deriveStrategyPnl` records that reading
 * NinjaTrader's clock column as a string puts "10:00 AM" before "9:35 AM", which
 * reorders an entire morning. The same applies to an epoch and to an order id -
 * "100" sorts before "99" - and a reordered book hands the first trade's targets
 * to the second. Text still compares as text, so a caller passing an ISO
 * timestamp is unaffected.
 */
function compare(left, right) {
  const a = Number(left);
  const b = Number(right);
  if (Number.isFinite(a) && Number.isFinite(b) && String(left).trim() !== '' && String(right).trim() !== '') {
    return a - b;
  }
  return String(left ?? '').localeCompare(String(right ?? ''));
}

function hasRoomFor(trade, order) {
  if (isStop(order)) return !trade.exits.some(isStop);
  if (!isTarget(order)) return false;
  const rung = targetNumber(order);
  return !trade.exits.some((exit) => isTarget(exit) && targetNumber(exit) === rung);
}

/**
 * The geometry a trade exhibits, in the units a template declares.
 *
 * Null when the entry never filled or the instrument has no tick size: a
 * distance measured against nothing is not a measurement.
 */
export function tradeGeometry(trade) {
  const entry = trade?.entry;
  if (!entry) return null;
  const tick = tickSizeFor(entry.instrument);
  if (!tick) return null;
  const basis = priceOf(entry.avgPrice) ?? priceOf(entry.limitPrice) ?? priceOf(entry.stopPrice);
  if (basis === null) return null;

  let stopTicks = 0;
  const rungs = new Map();
  for (const exit of trade.exits) {
    const price = priceOf(exit.limitPrice) ?? priceOf(exit.stopPrice);
    if (price === null) continue;
    const ticks = Math.round(Math.abs(price - basis) / tick);
    if (isStop(exit)) {
      if (stopTicks === 0) stopTicks = ticks;
    } else if (isTarget(exit)) {
      const rung = targetNumber(exit);
      if (rung >= 1 && rung <= 3 && !rungs.has(rung)) {
        rungs.set(rung, { ticks, size: Math.round(Number(exit.quantity) || 0) });
      }
    }
  }
  if (!rungs.size && !stopTicks) return null;
  return { instrument: instrumentRoot(entry.instrument), stopTicks, rungs };
}

/**
 * A zero is not a price.
 *
 * An order row carries both a limit and a stop column and the one it does not
 * use holds 0 rather than null. Taking the zero measures every stop from the
 * instrument's own price: on crude that produced stops 6,411 ticks away and
 * matched nothing at all. Filtering them took the same code from 0% to 55%.
 */
function priceOf(value) {
  const price = Number(value);
  return Number.isFinite(price) && price > 0 ? price : null;
}

/**
 * Which template a geometry belongs to.
 *
 * MATCHED ON WHAT THE TRADE SHOWS, NOT ON THE WHOLE TEMPLATE. Requiring every
 * declared rung meant only trades that ran all the way to a third target could
 * match - 446 of 4,782 on a real machine, and 5% of seven months recognised.
 * Comparing only the rungs actually placed takes it to 55%.
 *
 * Never fuzzy: versions are separated by as little as five ticks, so a
 * tolerance wide enough to absorb noise is wide enough to answer the wrong
 * version.
 */
export function matchGeometry(geometry, templates = []) {
  if (!geometry) return null;
  const candidates = (templates || []).filter((template) => fits(template, geometry));
  if (!candidates.length) return null;

  const families = [...new Set(candidates.map((template) => template.family))];
  if (families.length !== 1) return null;

  // The stop NARROWS, it does not reject - see `fits`. Where the observed stop
  // does pick out a subset, that subset decides the version; where it picks out
  // nothing, the targets answer on their own and the version may come back null.
  const exact = geometry.stopTicks > 0
    ? candidates.filter((template) => Number(template.stopTicks) === geometry.stopTicks)
    : [];
  const deciding = exact.length ? exact : candidates;

  const versions = [...new Set(deciding.map((t) => t.version).filter(Boolean))];
  const risks = [...new Set(deciding.map((t) => t.risk).filter(Boolean))];
  return {
    family: families[0],
    version: versions.length === 1 ? versions[0] : null,
    risk: risks.length === 1 ? risks[0] : null,
    candidates: candidates.length,
  };
}

/**
 * THE TARGETS DECIDE; THE STOP IS NOT ALLOWED TO REFUSE.
 *
 * A template declares where the stop was PLACED. NinjaTrader rewrites that
 * price as a trailing stop moves, so what the CRM stores is where the stop
 * ENDED, and comparing the two rejects a correct match for a reason that is not
 * a disagreement. Measured over seven months: 673 trades, 14% of the book,
 * failed on the stop alone.
 *
 * Letting it go costs nothing, and that is measured rather than assumed. Across
 * the desk's whole library exactly 3 template groups share their targets and
 * differ in their stop, and NONE of the three crosses families - every one is
 * two versions of the same algorithm. So the stop has never been what separates
 * one algorithm from another here; it separates versions, which is why
 * `matchGeometry` still uses it to pick the version and reports a null version
 * rather than a guess when it cannot.
 *
 * The agent's own copy keeps the strict comparison, because on the machine it
 * can read the original price out of OrderUpdates and has nothing to forgive.
 */
function fits(template, geometry) {
  if (!template) return false;
  if (instrumentRoot(template.instrument) !== geometry.instrument) return false;
  for (const [rung, observed] of geometry.rungs) {
    const ticks = Number(template[`target${rung}Ticks`]);
    const size = Number(template[`size${rung}`]);
    if (ticks !== observed.ticks || size !== observed.size) return false;
  }
  // A trade that showed no targets at all has only its stop to go on, and a
  // stop alone is not enough to name an algorithm.
  return geometry.rungs.size > 0;
}

/**
 * Attribute every order of one account.
 *
 * RECORD ALWAYS BEATS INFERENCE. An order that already carries a strategy name
 * is what the platform itself asserted, and nothing worked out here replaces
 * it. Everything else is the geometry's answer, or nothing.
 */
export function attributeOrders(orders = [], templates = []) {
  const results = [];
  for (const trade of reconstructTrades(orders)) {
    const match = matchGeometry(tradeGeometry(trade), templates);
    for (const order of [trade.entry, ...trade.exits]) {
      const recorded = String(order.strategyName || '').trim();
      if (recorded) {
        results.push({ id: order.id, family: recorded, version: null, basis: 'record' });
      } else if (match) {
        results.push({ id: order.id, family: match.family, version: match.version, basis: 'inferred' });
      }
      // No row at all when neither settles it: absence is the honest answer and
      // the column stays null, which is what "nobody has worked this out" looks
      // like on every screen that reads it.
    }
  }
  return results;
}
